@jules address issue and merge all changes back to this pr: Run bash scripts/validate.sh
==> cargo fmt --check
==> cargo clippy --all-targets --all-features -- -D warnings
    Finished `dev` profile [unoptimized] target(s) in 0.16s
==> cargo test --no-run --all-features (check for warnings)
==> cargo test --all-targets
   Compiling chaotic_semantic_memory v0.3.4 (/home/runner/work/chaotic_semantic_memory/chaotic_semantic_memory)
    Finished `test` profile [unoptimized] target(s) in 11.46s
     Running unittests src/lib.rs (target/debug/deps/chaotic_semantic_memory-dd75c00774d5f710)

running 111 tests
test bridge_retrieval::tests::test_bridge_retrieval_empty_graph ... ok
test bridge_retrieval::tests::test_bridge_retrieval_empty_singularity ... ok
test bridge_retrieval::tests::test_bridge_retrieval_with_expansion ... ok
test bridge_retrieval::tests::test_final_score_weights ... ok
test bridge_retrieval::tests::test_memory_packet_empty_hits ... ok
test bundle::tests::test_bundle_accumulator_add_finalize ... ok
test bundle::tests::test_bundle_accumulator_clear ... ok
test bundle::tests::test_bundle_accumulator_empty ... ok
test bundle::tests::test_bundle_accumulator_remove ... ok
test cli::commands::index_dir::tests::test_chunk_by_heading_basic ... ok
test cli::commands::index_dir::tests::test_chunk_by_heading_empty ... ok
test cli::commands::index_dir::tests::test_chunk_by_heading_min_level ... ok
test cli::commands::index_dir::tests::test_parse_heading_invalid ... ok
test cli::commands::index_dir::tests::test_parse_heading_valid ... ok
test cli::error::tests::test_exit_code_mapping ... ok
test cli::error::tests::test_memory_error_conversion ... ok
test cli::git_local::tests::test_ensure_git_local_dir ... ok
test cli::git_local::tests::test_resolve_git_local_path_in_git_repo ... ok
test cli::git_local::tests::test_resolve_git_local_path_outside_git_repo ... ok
test cli::git_local::tests::test_resolve_git_local_path_structure ... ok
test concept_builder::tests::concept_builder_creates_concept_with_metadata ... ok
test concept_builder::tests::concept_builder_uses_random_vector_by_default ... ok
test concept_builder::tests::concept_builder_with_ttl_sets_expiration ... ok
test concept_builder::tests::concept_builder_without_ttl_has_no_expiration ... ok
test error::tests::database_error_exposes_source_chain ... ok
test error::tests::reservoir_error_exposes_source_chain ... ok
test framework_bridge::tests::test_memory_packet_text_empty ... ok
test framework_bridge::tests::test_probe_bridge_text_empty ... ok
test bridge_persistence::tests::test_save_and_load_canonical_concept ... ok
test framework_validation::tests::test_validate_concept_id_dangerous_chars ... ok
test framework_validation::tests::test_validate_concept_id_empty ... ok
test framework_validation::tests::test_validate_concept_id_too_long ... ok
test framework_validation::tests::test_validate_retrieval_config_bucket_width ... ok
test framework_validation::tests::test_validate_traversal_config ... ok
test framework_bridge::tests::test_probe_bridge_text_with_concepts ... ok
test graph_traversal::tests::test_bfs_max_depth ... ok
test bridge_persistence::tests::test_save_and_load_concept_graph ... FAILED
test graph_traversal::tests::test_bfs_missing_concept ... ok
test graph_traversal::tests::test_bfs_simple ... ok
test graph_traversal::tests::test_incoming_associations ... ok
test graph_traversal::tests::test_neighbors ... ok
test graph_traversal::tests::test_shortest_path_direct ... ok
test graph_traversal::tests::test_shortest_path_dijkstra_prefers_strong_edge ... ok
test graph_traversal::tests::test_shortest_path_hops_indirect ... ok
test graph_traversal::tests::test_shortest_path_hops_direct ... ok
test graph_traversal::tests::test_shortest_path_hops_no_path ... ok
test bridge_persistence::tests::test_delete_canonical_concept ... ok
test graph_traversal::tests::test_shortest_path_indirect ... ok
test graph_traversal::tests::test_shortest_path_no_path ... ok
test graph_traversal::tests::test_shortest_path_same_node ... ok
test hyperdim::tests::test_hvec_creation ... ok
test hyperdim::tests::test_binding ... ok
test hyperdim::tests::test_permute ... ok
test hyperdim::tests::test_json_serialize_is_base64 ... ok
test hyperdim::tests::test_random_generation ... ok
test hyperdim::tests::test_serialization ... ok
test hyperdim::tests::test_json_array_deserialize_fallback ... ok
test hyperdim::tests::test_self_similarity ... ok
test hyperdim_batch::tests::empty_batch_returns_empty_vec ... ok
test hyperdim_batch::tests::batch_matches_scalar_results ... ok
test metadata_filter::tests::test_and_one_fails ... ok
test metadata_filter::tests::test_and_all_match ... ok
test metadata_filter::tests::test_eq_match ... ok
test metadata_filter::tests::test_eq_missing_key ... ok
test metadata_filter::tests::test_exists_missing ... ok
test metadata_filter::tests::test_eq_no_match ... ok
test metadata_filter::tests::test_in_match ... ok
test metadata_filter::tests::test_in_no_match ... ok
test metadata_filter::tests::test_nested_complex ... ok
test metadata_filter::tests::test_exists_present ... ok
test metadata_filter::tests::test_not ... ok
test metadata_filter::tests::test_or_any_match ... ok
test metadata_filter::tests::test_or_none_match ... ok
test retrieval::bm25::tests::test_add_document ... ok
test retrieval::bm25::tests::test_avg_doc_length ... ok
test retrieval::bm25::tests::test_custom_config ... ok
test retrieval::bm25::tests::test_clear ... ok
test retrieval::bm25::tests::test_no_matching_terms ... ok
test retrieval::bm25::tests::test_remove_document ... ok
test retrieval::bm25::tests::test_idf_rare_term_higher_score ... ok
test retrieval::bm25::tests::test_replace_document ... ok
test retrieval::bm25::tests::test_doc_length_normalization ... ok
test retrieval::bm25::tests::test_search_empty_index ... ok
test retrieval::bm25::tests::test_search_empty_query ... ok
test retrieval::bm25::tests::test_search_exact_match ... ok
test retrieval::bm25::tests::test_single_term_query ... ok
test retrieval::bm25::tests::test_search_partial_match ... ok
test retrieval::bm25::tests::test_top_k_zero_returns_empty ... ok
test retrieval::hybrid::tests::test_compute_weights_long_query ... ok
test retrieval::bm25::tests::test_zero_length_document ... ok
test retrieval::bm25::tests::test_top_k ... ok
test retrieval::hybrid::tests::test_compute_weights_medium_query ... ok
test retrieval::hybrid::tests::test_compute_weights_short_query ... ok
test retrieval::hybrid::tests::test_compute_weights_very_long_query ... ok
test retrieval::hybrid::tests::test_merge_results_basic ... ok
test hyperdim::tests::test_bundle ... ok
test retrieval::hybrid::tests::test_merge_results_empty ... ok
test retrieval::hybrid::tests::test_merge_results_weighted ... ok
test retrieval::hybrid::tests::test_normalize_scores_equal ... ok
test retrieval::hybrid::tests::test_normalize_scores_basic ... ok
test semantic_bridge::tests::test_bridge_config_defaults ... ok
test retrieval::hybrid::tests::test_normalize_scores_empty ... ok
test semantic_bridge::tests::test_canonical_concept_builder ... ok
test semantic_bridge::tests::test_concept_graph_expand ... ok
test semantic_bridge::tests::test_concept_graph_add_and_get ... ok
test semantic_bridge::tests::test_concept_graph_expand_no_cycle ... ok
test semantic_bridge::tests::test_concept_graph_remove ... ok
test semantic_bridge::tests::test_concept_graph_match_tokens_case_insensitive ... ok
error: test failed, to rerun pass `--lib`
test semantic_triples::tests::test_triple_formatting ... ok
test semantic_bridge::tests::test_memory_packet_estimated_tokens ... ok
test semantic_triples::tests::test_payload_formatting ... ok

failures:

---- bridge_persistence::tests::test_save_and_load_concept_graph stdout ----

thread 'bridge_persistence::tests::test_save_and_load_concept_graph' (4137) panicked at src/bridge_persistence.rs:300:54:
called `Result::unwrap()` on an `Err` value: Database { message: "Failed to save canonical concept: SQLite failure: `UNIQUE constraint failed: csm_canonical.id`", source: None }
stack backtrace:
   0: __rustc::rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::result::unwrap_failed
   3: chaotic_semantic_memory::bridge_persistence::tests::test_save_and_load_concept_graph::{{closure}}
   4: <core::pin::Pin<P> as core::future::future::Future>::poll
   5: <core::pin::Pin<P> as core::future::future::Future>::poll
   6: tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}::{{closure}}::{{closure}}
   7: tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}::{{closure}}
   8: tokio::runtime::scheduler::current_thread::Context::enter
   9: tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}
  10: tokio::runtime::scheduler::current_thread::CoreGuard::enter::{{closure}}
  11: tokio::runtime::context::scoped::Scoped<T>::set
  12: tokio::runtime::context::set_scheduler::{{closure}}
  13: std::thread::local::LocalKey<T>::try_with
  14: std::thread::local::LocalKey<T>::with
  15: tokio::runtime::context::set_scheduler
  16: tokio::runtime::scheduler::current_thread::CoreGuard::enter
  17: tokio::runtime::scheduler::current_thread::CoreGuard::block_on
  18: tokio::runtime::scheduler::current_thread::CurrentThread::block_on::{{closure}}
  19: tokio::runtime::context::runtime::enter_runtime
  20: tokio::runtime::scheduler::current_thread::CurrentThread::block_on
  21: tokio::runtime::runtime::Runtime::block_on_inner
  22: tokio::runtime::runtime::Runtime::block_on
  23: chaotic_semantic_memory::bridge_persistence::tests::test_save_and_load_concept_graph
  24: chaotic_semantic_memory::bridge_persistence::tests::test_save_and_load_concept_graph::{{closure}}
  25: core::ops::function::FnOnce::call_once
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.


failures:
    bridge_persistence::tests::test_save_and_load_concept_graph

test result: FAILED. 110 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.03s

Error: Process completed with exit code 101.

--- a/src/bridge_persistence.rs
+++ b/src/bridge_persistence.rs
@@ -148,7 +148,7 @@ impl Persistence {
             .map_err(|e| MemoryError::database(format!("Failed to begin transaction: {}", e)))?;
 
         // Clear existing concepts
-        if let Err(e) = conn.execute("DELETE FROM csm_canonical", params![]).await {
+        if let Err(e) = conn.execute("DELETE FROM csm_canonical", ()).await {
             let _ = conn.execute("ROLLBACK", ()).await;
             return Err(MemoryError::database(format!(
                 "Failed to clear canonical concepts: {}",
@@ -156,7 +156,7 @@ impl Persistence {
             )));
         }
 
-        // Insert all concepts
+        // Insert all concepts with prepared statement for efficiency
         let mut first_error: Option<MemoryError> = None;
         let stmt = match conn
             .prepare(
@@ -190,6 +190,9 @@ impl Persistence {
                     break;
                 }
             };
+
+            // Reset statement before each execution for proper parameter binding
+            stmt.reset();
 
             if let Err(e) = stmt
                 .execute(params![

